ecosystem/typescript/sdk/src/transaction_builder/bcs/serializer.ts

```diff
@@ -0,0 +1,215 @@
+    }
+    const byteValue = value ? 1 : 0;
+    this.serialize(new Uint8Array([byteValue]));
+  }
+
+  /**
+   * Serializes a uint8 number.
+   *
+   * BCS layout for "uint8": One byte. Binary format in little-endian representation.
+   */
+  @checkNumberRange(0, MAX_U8_NUMBER)
+  serializeU8(value: Uint8): void {
+    this.serialize(new Uint8Array([value]));
+  }
+
+  /**
+   * Serializes a uint16 number.
+   *
+   * BCS layout for "uint16": Two bytes. Binary format in little-endian representation.
+   * @example
+   * ```ts
+   * const serializer = new Serializer();
+   * serializer.serializeU16(4660);
+   * assert(serializer.getBytes() === new Uint8Array([0x34, 0x12]));
+   * ```
+   */
+  @checkNumberRange(0, MAX_U16_NUMBER)
+  serializeU16(value: Uint16): void {
+    this.serializeWithFunction(DataView.prototype.setUint16, 2, value);
+  }
+
+  /**
+   * Serializes a uint32 number.
+   *
+   * BCS layout for "uint32": Four bytes. Binary format in little-endian representation.
+   * @example
+   * ```ts
+   * const serializer = new Serializer();
+   * serializer.serializeU32(305419896);
+   * assert(serializer.getBytes() === new Uint8Array([0x78, 0x56, 0x34, 0x12]));
+   * ```
+   */
+  @checkNumberRange(0, MAX_U32_NUMBER)
+  serializeU32(value: Uint32): void {
+    this.serializeWithFunction(DataView.prototype.setUint32, 4, value);
+  }
+
+  /**
+   * Serializes a uint64 number.
+   *
+   * BCS layout for "uint64": Eight bytes. Binary format in little-endian representation.
+   * @example
+   * ```ts
+   * const serializer = new Serializer();
+   * serializer.serializeU64(1311768467750121216);
+   * assert(serializer.getBytes() === new Uint8Array([0x00, 0xEF, 0xCD, 0xAB, 0x78, 0x56, 0x34, 0x12]));
+   * ```
+   */
+  @checkNumberRange(0n, MAX_U64_BIG_INT)
+  serializeU64(value: AnyNumber): void {
+    const low = BigInt(value.toString()) & BigInt(MAX_U32_NUMBER);
+    const high = BigInt(value.toString()) >> BigInt(32);
+
+    // write little endian number
+    this.serializeU32(Number(low));
+    this.serializeU32(Number(high));
+  }
+
+  /**
+   * Serializes a uint128 number.
+   *
+   * BCS layout for "uint128": Sixteen bytes. Binary format in little-endian representation.
+   */
+  @checkNumberRange(0n, MAX_U128_BIG_INT)
+  serializeU128(value: AnyNumber): void {
+    const low = BigInt(value.toString()) & MAX_U64_BIG_INT;
+    const high = BigInt(value.toString()) >> BigInt(64);
+
+    // write little endian number
+    this.serializeU64(low);
+    this.serializeU64(high);
+  }
+
+  /**
+   * Serializes a uint32 number with uleb128.
+   *
+   * BCS use uleb128 encoding in two cases: (1) lengths of variable-length sequences and (2) tags of enum values
+   */
+  @checkNumberRange(0, MAX_U32_NUMBER)
+  serializeU32AsUleb128(val: Uint32): void {
+    let value = val;
+    const valueArray = [];
+    while (value >>> 7 !== 0) {
+      valueArray.push((value & 0x7f) | 0x80);
+      value >>>= 7;
+    }
```

Comment:
Am I missing something, I'm not sure what this is supposed to be doing? It seems to be encoding 7 bits and always adding the most significant bit.